### ISSUE TITLE
Hyperparameter optimization with hyperopt

### ISSUE THREAD

**Issue by osma** [NatLibFi/Annif#240]
Some backends require preselected (hyper)parameters and these must be carefully chosen to get the best results. Often the only way is trial and error. Especially fastText is notorious for requiring a large number of parameters and finding the magic combination can be a challenge.

The [hyperopt](https://github.com/hyperopt/hyperopt) library implements an algorithm for efficiently searching through the hyperparameter space in order to find a good combination. We could implement a command like `annif hyperopt <project> <corpus>` that looks for good hyperparameters.

**Comment by osma** [comment]
Here is a snippet I've used previously to find suitable fastText hyperparameters. It requires training and validation files in the fastText format.

```python
from hyperopt import fmin, tpe, hp, Trials
import sys
import fastText

trainfile = sys.argv[1]
validfile = sys.argv[2]
evals = int(sys.argv[3])

trials = Trials()

# precision@K value
K_VALUE = 5

def eval(space):
    print("Trial {} of {}".format(len(trials), evals))
    print("Testing values:", space)
    model = fastText.train_supervised(
        input=trainfile,
        dim=int(space['dim']),
        lr=space['lr'],
        epoch=int(space['epoch']),
        minCount=int(space['minCount']),
        wordNgrams=int(space['ngrams']),
        minn=int(space['minn']),
        maxn=int(space['maxn']),
        loss='hs'
    )
    n, p, r = model.test(validfile, k=K_VALUE)
    print("Precision:", p)
    print()
    return 1-p

space = {
    'dim': hp.quniform('dim', 100, 500, 10),
    'lr': hp.loguniform('lr', -0.4, -0.1),
    'epoch': hp.quniform('epoch', 20, 120, 5),
    'minCount': hp.quniform('minCount', 1, 5, 1),
    'ngrams': hp.choice('ngrams', [1, 2]),
    'minn': hp.quniform('minn', 3, 5, 1),
    'maxn': hp.quniform('maxn', 5, 7, 1)
}

best = fmin(
    fn=eval,
    space=space,
    algo=tpe.suggest,
    max_evals=evals,
    trials=trials
)

print("Best parameters:", best)
print("Best precision:", 1 - trials.best_trial['result']['loss'])
```

**Comment by mo-fu** [comment]
Two other alternative libraries for hyper parameter optimization:
* https://github.com/dragonfly/dragonfly
*  https://github.com/automl/SMAC3

**Comment by osma** [comment]
The optimization of ensemble weights (#321) was implemented in PR #414, just merged for 0.49. But other kinds of hyperparameter optimization (e.g. searching for optimal fasttext and omikuji hyperparameters) are left for later.

**Comment by osma** [comment]
I've opened backend-specific issues for the remaining cases where hyperparameter optimization would make sense. Closing this issue, since the core support for hyperparameter optimization has already been implemented for 0.49.
